transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -0,0 +1,355 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.android.datatransport.runtime.scheduling.persistence;
+
+import android.content.ContentValues;
+import android.content.Context;
+import android.database.Cursor;
+import android.database.sqlite.SQLiteDatabase;
+import android.database.sqlite.SQLiteOpenHelper;
+import android.support.annotation.Nullable;
+import com.google.android.datatransport.Priority;
+import com.google.android.datatransport.runtime.EventInternal;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Set;
+import javax.inject.Inject;
+
+/** {@link EventStore} implementation backed by a SQLite database. */
+public class SQLiteEventStore implements EventStore {
+
+  static final int MAX_RETRIES = 10;
+
+  private static final Priority[] ALL_PRIORITIES = Priority.values();
+
+  private final OpenHelper openHelper;
+  private SQLiteDatabase db;
+
+  @Inject
+  SQLiteEventStore(Context applicationContext) {
+    this.openHelper = new OpenHelper(applicationContext);
+  }
+
+  private SQLiteDatabase getDb() {
+    if (db == null) {
+      db = openHelper.getWritableDatabase();
```

Comment:
Your in-memory EventStore synchronizes many methods leading me to believe you're defending against multiple threads storing data this way.

This assignment is unsynchronized, which seems like you could end up with multiple opens happening concurrently, which possibly could result in multiple OpenHelpers attempting to create the schema concurrently.

I don't think making `getDb()` synchronized would be enough, because your implementation of atomically essentially assumes that `db` has a current transaction in force. Multiple concurrent callers would attempt to use that same transaction from multiple threads.

Reply:
Good catch, but it is not the case. We are explicitly designing the library to be single-threaded as it greatly simplifies the design. Essentially all db access will happen on a single thread sequentially.

What's not entirely clear to me is how my setup will behave in a multiprocess setup which we have to plan for being a low-level library that's going to be used by other SDKs. Specifically, below I issue `PRAGMA busy_timeout = 5000` in the hope that concurrent access to the db file won't immediately result in `sqlite busy` but would instead wait for the transaction lock to get released and then proceed normally. Does this approach seem reasonable to you?

Reply:
That's not clear to me.

Within Firestore we tie which SQLite DB we open to the current FirebaseApp and then open then exclusively. We disallow multiple processes to access the same DB concurrently.

The mechanism you're using isn't one I'm familiar with.

Reply:
OK, so just to document my findings.

`PRAGMA busy_timeout` is only supported in sqlite `3.7.15+`, which is on present only api21+ devices so we can't use that, instead I will implement blocking/waiting manually.